src/app/modules/item/task/task-xd-pr.ts

```diff
@@ -0,0 +1,303 @@
+  minScore: number,
+  maxScore: number,
+  noScore: number,
+  randomSeed: number,
+  readOnly: boolean,
+  options: Object,
+}
+export type TaskParamsValue = TaskParams | Object | string | number | undefined;
+
+export interface UpdateDisplayParams {
+  height?: number | string,
+  views?: Object,
+  scrollTop?: number,
+}
+
+// TODO : actual types
+export type TaskMetaData = any;
+export type TaskView = any;
+export type TaskViews = any;
+export type TaskGrade = any;
+export type TaskResources = any;
+export type TaskDisplayData = any;
+export type TaskLog = any;
+
+function getUrlParameterByName(name : string, url : string) : string {
+  const regex = new RegExp("[\\?&]" + name + "=([^&#]*)"),
+    results = regex.exec(url);
+  return results && results[1] ? decodeURIComponent(results[1].replace(/\+/g, " ")) : "";
+}
+
+function getRandomID(): string {
+  const low = Math.floor(Math.random() * 922337203).toString();
+  const high = Math.floor(Math.random() * 2000000000).toString();
+  return high + low;
+}
+
+/** Get URL for a task with specific parameters */
+export function getTaskUrl(taskUrl : string, token : string, platform : string, prefix? : string, locale? : string) : string {
+  const channelId = (prefix || '') + getRandomID();
+  if (taskUrl.indexOf('?') == -1) {
+    // the idea is not to change the base url even if we change token, so we put token after #
+    taskUrl = taskUrl + '?';
+  } else {
+    taskUrl = taskUrl + '&';
+  }
+  let url = taskUrl
+      + 'sToken=' + encodeURIComponent(token)
+      + '&sPlatform=' + encodeURIComponent(platform)
+      + '&channelId=' + encodeURIComponent(channelId);
+  if (locale) {
+    url += '&sLocale=' + encodeURIComponent(locale);
+  }
+  return url;
+}
+
+/** Get Task object from an iframe */
+export function getTaskProxy(iframe : HTMLIFrameElement): Observable<Task> {
+  // Check every second whether the iframe loaded
+  return interval(1000)
+    .pipe(
+      // Abort after 300 seconds
+      map(
+        n => {
+          if (n > 300) {
+            throw Error("IFrame load waiting timeout");
+          }
+          return n;
+        }
+      ),
+
+      // Get contentWindow when it exists
+      filter(_ => !!iframe.contentWindow),
+      take(1),
+      map(_ => iframe.contentWindow as Window),
+
+      // Get the RxMessagingChannel from the contentWindow
+      switchMap((contentWindow: Window) =>
+        rxBuild({
+          window: contentWindow,
+          origin: "*",
+          scope: getUrlParameterByName('channelId', iframe.src),
+        })
+      ),
+
+      // Get the Task from the channel
+      map((chan: RxMessagingChannel) => new Task(chan))
+    );
+}
+
+
+/**
+ * Task object, created from an iframe DOM element
+ */
+export class Task {
+  platformSet = false;;
+  chan: RxMessagingChannel;
+
+  constructor(chan: RxMessagingChannel) {
+    this.chan = chan;
+  }
+
+  destroy(): void {
+    this.chan.destroy();
+  }
+
+  setPlatform(platform : Platform): void {
```

Comment:
bindPlatform/bindListener (if platform has been renamed to listener)

Reply:
Renamed to `bindPlatform`.